### PR TITLE
Adopt PyTorch official LRScheduler API

### PR DESCRIPTION
**Pull request by tgaddair**

### CODE REVIEW

File: ludwig/modules/lr_scheduler.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+class ReduceLROnPlateauLimited(ReduceLROnPlateau):
+    def __init__(self, optimizer: Optimizer, mode: str, reduce_limit: int, factor: float, patience: int):
+        super().__init__(optimizer, mode=mode, factor=factor, patience=patience)
+        self.reduce_limit = reduce_limit
+        self._num_reduce_lr = 0
+
+    def step(self, metrics, epoch=None):
+        if self._num_reduce_lr >= self.reduce_limit:
+            # Already reduce the LR as many times as we will allow
```
Comment (justinxzhao) [inline ludwig/modules/lr_scheduler.py:23]: "reduce" --> "reduced"

File: ludwig/schema/lr_scheduler.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+
+# TODO(travis): too much boilerplate here, we should find a way to abstract all this and only require specifying the
+# minimal amount needed for the new config object.
+@DeveloperAPI
+def LRSchedulerDataclassField(description: str, default: Dict = None):
+    """Returns custom dataclass field for `LRSchedulerConfig`. Allows `None` by default.
+
+    :param description: Description of the dataclass field
+    :param default: dict that specifies param values that will be loaded by its schema class (default: None).
```
Comment (justinxzhao) [inline ludwig/schema/lr_scheduler.py:106]: nit: Google-style args docstrings.

File: ludwig/modules/lr_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        self.config = config
+        self.reset(steps_per_checkpoint, total_steps)
+
+    def reset(self, steps_per_checkpoint: int, total_steps: int):
+        self.steps_per_checkpoint = steps_per_checkpoint
+        self.num_training_steps = total_steps
+
+        if self.config.warmup_fraction > 0 and self.config.warmup_evaluations > 0:
+            logging.info(
+                "Both `learning_rate_scheduler.warmup_fraction` and `learning_rate_scheduler.warmup_evaluations`. "
```
Comment (justinxzhao) [inline ludwig/modules/lr_scheduler.py:111]: nit: Both warmups "are specified" or "are non-zero".

"The larger of the two (as a function of the total training steps) will be used."

File: ludwig/modules/lr_scheduler.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+            "eval_scheduler_state": self._eval_scheduler.state_dict() if self._eval_scheduler is not None else {},
+        }
+
+    def load_state_dict(self, d: Dict[str, Any]):
+        self._train_scheduler.load_state_dict(d["train_scheduler_state"])
+        if self._eval_scheduler is not None:
+            self._eval_scheduler.load_state_dict(d["eval_scheduler_state"])
+
+
+class StepInfo:
```
Comment (justinxzhao) [inline ludwig/modules/lr_scheduler.py:100]: nit: Add a docstring.

File: ludwig/modules/lr_scheduler.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        if self._eval_scheduler is None:
+            # No reduce on plateau
+            return
+
+        if self.config.reduce_eval_split == TRAINING:
+            split_metrics = progress_tracker.train_metrics
+        elif self.config.reduce_eval_split == VALIDATION:
+            split_metrics = progress_tracker.validation_metrics
+        else:  # if self.config.reduce_eval_split == TEST:
+            split_metrics = progress_tracker.test_metrics
```
Comment (justinxzhao) [inline ludwig/modules/lr_scheduler.py:80]: Not in scope for this PR, but we could/should be more prescriptive about which split to use for reduction (training or validation). Using `TEST` doesn't seem quite right. CC: @dantreiman @w4nderlust

File: ludwig/trainers/trainer.py [hunk 6]
```diff
@@ -1087,31 +1032,8 @@ def check_progress_on_validation(
...
-                and progress_tracker.last_reduce_learning_rate_eval_metric_improvement > 0
-                and not progress_tracker.num_reductions_learning_rate >= reduce_learning_rate_on_plateau
-            ):
-                logger.info(
-                    f"Last learning rate reduction happened {progress_tracker.last_learning_rate_reduction} step(s) "
-                    f"ago, improvement of {validation_output_feature_name} {reduce_learning_rate_eval_split} "
-                    f"{reduce_learning_rate_eval_metric} happened "
-                    f"{progress_tracker.last_reduce_learning_rate_eval_metric_improvement} step(s) ago."
-                )
+        # ========== Learning Rate Schedule evaluation updates ========
```
Comment (justinxzhao) [inline ludwig/trainers/trainer.py:1035]: This simplification is awesome.

File: ludwig/trainers/trainer.py [hunk 7]
```diff
@@ -181,19 +163,26 @@ def __init__(
         if self.device is None:
             self.device = get_torch_device()
 
+        base_learning_rate = config.learning_rate
+        if self.horovod:
```
Comment (justinxzhao) [inline ludwig/trainers/trainer.py:167]: This will become self.distributed after  https://github.com/ludwig-ai/ludwig/pull/2890/files lands, right?

File: ludwig/modules/lr_scheduler.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+from ludwig.utils.trainer_utils import ProgressTracker
+
+
+class ReduceLROnPlateauLimited(ReduceLROnPlateau):
+    def __init__(self, optimizer: Optimizer, mode: str, reduce_limit: int, factor: float, patience: int):
+        super().__init__(optimizer, mode=mode, factor=factor, patience=patience)
+        self.reduce_limit = reduce_limit
+        self._num_reduce_lr = 0
+
+    def step(self, metrics, epoch=None):
```
Comment (justinxzhao) [inline ludwig/modules/lr_scheduler.py:21]: It looks like `epoch` is deprecated. Let's exclude it: https://pytorch.org/docs/stable/_modules/torch/optim/lr_scheduler.html#ReduceLROnPlateau

File: ludwig/modules/lr_scheduler.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+from torch.optim.lr_scheduler import LambdaLR, ReduceLROnPlateau
+
+from ludwig.constants import MINIMIZE, TRAINING, VALIDATION
+from ludwig.modules.metric_modules import get_metric_cls
+from ludwig.schema.lr_scheduler import LRSchedulerConfig
+from ludwig.utils.metric_utils import TrainerMetric
+from ludwig.utils.trainer_utils import ProgressTracker
+
+
+class ReduceLROnPlateauLimited(ReduceLROnPlateau):
```
Comment (justinxzhao) [inline ludwig/modules/lr_scheduler.py:15]: Alternative name: `ReduceLROnPLateauCappedDecreases`.

**Review by justinxzhao — COMMENTED** [review]
Nice changes! Overall, really like the simplification of `trainer.py`.

File: ludwig/modules/lr_scheduler.py (hunk 5 above)
Comment (tgaddair) [reply ludwig/modules/lr_scheduler.py:80]: Agreed test seems like a violation of the concept of a test set. I think training or validation both have merits, but usually you'd want to use training, I'd say.

File: ludwig/trainers/trainer.py (hunk 7 above)
Comment (tgaddair) [reply ludwig/trainers/trainer.py:167]: Yes.

**Review by justinxzhao — APPROVED** [review]
LGTM
